### DISCUSSION TITLE
FNS and FUNCTION; VARS and VARIABLES

### DISCUSSION THREAD

**Discussion by masinter** [Interlisp/medley#1162] [Yearnings]
I think I remember why we approached this from a "don't mess up Interlisp" point of view, but the result is ugly and confusing and buggy. 

For example, there is a lot of patching around problems when there is a FNS definition of a function on one file and a FUNCTIONS definition on the other file.

FNS goes into the symbol definition cell and VARS goes into the value pointer, while FUNCTIONS and VARIABLES are linked to a "definer" expression. 
* There's a problem in ED and other interfaces which are trying to compute HASDEF. I think the underlying mechanism needs a way of saying "subtype" so that FNS is a sub-type of FUNCTIONS rather than a separate type. If something HASDEF VARIABLES then it HASDEF VARS too, as with FUNCTIONS and FNS and the code needs to be aware of that relationship... TYPESOF(symbol) should return FNS if it only has FNS but FUNCTIONS if it has any FUNCTIONS anywhere?

**Comment by masinter** [comment]
I think it is possible and would be helpful to eliminate the difference between FNS and FUNCTIONS -- make them synonyms. Add an IL:DEFUN and use it instead of DEFINEQ. The distinction between VARS and VARIABLES is a little trickier because of ALISTS and ADDVARS and INITVARS.

**Comment by nbriggs** [comment]
@masinter - I hope that whatever happens it includes updating the relevant documentation, before the change is released.

**Comment by pamoroso** [comment]
Thousands of pages worth of valuable documentation assume `DEFINEQ` and the distinction between `FNS` and `FUNCTIONS`.
